pkg/lib/cockpit-components-context-menu.jsx

```diff
@@ -0,0 +1,98 @@
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from "react";
+
+import "context-menu.css";
+
+export class ContextMenu extends React.Component {
+    constructor() {
+        super();
+        this.state = { visible: false };
+        this._handleContextMenu = this._handleContextMenu.bind(this);
+        this._handleClick = this._handleClick.bind(this);
+    }
+
+    componentDidMount() {
+        document.addEventListener('contextmenu', this._handleContextMenu);
+        document.addEventListener('click', this._handleClick);
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener('contextmenu', this._handleContextMenu);
+        document.removeEventListener('click', this._handleClick);
+    }
+
+    _handleContextMenu(event) {
+        event.preventDefault();
+
+        this.setState({ visible: true });
+
+        const clickX = event.clientX;
+        const clickY = event.clientY;
+        const screenW = window.innerWidth;
+        const screenH = window.innerHeight;
+        const rootW = this.root.offsetWidth;
+        const rootH = this.root.offsetHeight;
+
+        const right = (screenW - clickX) > rootW;
+        const left = !right;
+        const top = (screenH - clickY) > rootH;
+        const bottom = !top;
+
+        if (right) {
+            this.root.style.left = `${clickX + 5}px`;
+        }
+
+        if (left) {
+            this.root.style.left = `${clickX - rootW - 5}px`;
+        }
+
+        if (top) {
+            this.root.style.top = `${clickY + 5}px`;
+        }
+
+        if (bottom) {
+            this.root.style.top = `${clickY - rootH - 5}px`;
+        }
+    }
+
+    _handleClick(event) {
+        if (event && event.button === 0) {
+            const wasOutside = !(event.target.contains === this.root);
+
+            if (wasOutside && this.state.visible)
+                this.setState({ visible: false });
+        }
+    }
+
+    render() {
+        return this.state.visible &&
+            <div ref={ ref => { this.root = ref } } className="contextMenu">
+                <div className="contextMenuOption" onClick={this.props.getText}>
+                    <div className="contextMenuName">Copy</div>
+                    <div className="contextMenuShortcut">Ctrl+Insert</div>
+                </div>
+                <div className="contextMenuOption" onClick={this.props.setText} >
+                    <div className="contextMenuName">Paste</div>
+                    <div className="contextMenuShortcut">Shift+Insert</div>
+                </div>
+            </div>;
+    }
+}
```

Comment:
Please declare and document the properties of this component, especially because it is in pkg/lib.

Reply:
Declaring the properties still applies. Look at `git grep propTypes` for examples.

Reply:
aha, I misunderstood what you meant. Will fix that.